# HW02: Introduction to `Python` and `Pandas`

For this problem set, we will be doing some basic programming excercises in python to get you acquainted with using python and jupyter notebooks.

## Preambule
### Guidelines

All problem sets must be submitted to moodle a compiled *notebook* file in PDF. You can achieve this by exporting the notebook in PDF. 

### <span style='color:green'> Questions to answer are in green</span>

### Loading Packages

One of the things that makes python such a powerful programming+statistical tool are the freely available and high quality packages which are constantly being written by users almost daily. For our purposes we will be loading two packages: 

* `pandas` package for data management;
* `matplotlib` for plotting, and `seaborn` for nicer plots

In [1]:
import matplotlib.pyplot as plt
%matplotlib inline
import numpy as np
import pandas as pd
import seaborn as sn
pd.set_option('display.max_columns', 500)

### Background

We will be using the [Stop, Question and Frisk Data from the NYPD](https://www1.nyc.gov/site/nypd/stats/reports-analysis/stopfrisk.page) which contains information about over 100,000 police citizen interactions between 2003-2016 (but we use it for 2016 only).

Throughout the notebook, we study how individual level characteristics correlates with being arrested, aim at describing whether arrests are showing evidence of racism. 

We will explore the connection between the following variables:

* **arstmade** - Was an arrest made?
* **race** - Race of the suspect.
* **timestop** - Time that the suspect was stopped. 
* **datestop** - Date that the suspect was stopped. 
* **age** - Suspect's age.


## Importing and cleaning data

In [2]:
# Importation from a URL, but works with a filepath instead
df=pd.read_csv("https://www1.nyc.gov/assets/nypd/downloads/excel/analysis_and_planning/stop-question-frisk/sqf-2016.csv")

/home/guillotm/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (0,1,2,3,4,8,18,19,73,74,84,85,86,103,110,111) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Looking at the data

In [3]:
df.columns

Index(['year', 'pct', 'ser_num', 'datestop', 'timestop', 'recstat', 'inout',
       'trhsloc', 'perobs', 'crimsusp',
       ...
       'zip', 'addrpct', 'sector', 'beat', 'post', 'xcoord', 'ycoord',
       'dettypCM', 'lineCM', 'detailCM'],
      dtype='object', length=112)

In [4]:
df.shape

(12405, 112)

This tells us that we have 12,405 observations (stops) and have 112 variables which were collected for the stops.

In [5]:
# have a look at the first rows
df.head()

year pct ser_num datestop timestop recstat inout trhsloc perobs  \
0  2016  41      22  2072016      100       A     O       P      1   
1  2016  10      22  2182016       30       1     O       P      8   
2  2016  66       1  1012016       30       1     I       P      2   
3  2016  47      18  1012016       40       1     O       H      1   
4  2016  79       1  1012016       50       1     O       P      3   

      crimsusp perstop typeofid explnstp othpers arstmade arstoffn sumissue  \
0         BURG       8        P        Y       N        N                 Y   
1  MISDEMEANOR      10        P        Y       Y        N                 N   
2          FEL      10        P        Y       N        Y  ASSAULT        N   
3          FEL       2        V        Y       N        N                 N   
4       D.W.I.      10        P        Y       N        N                 N   

  sumoffen compyear comppct offunif officrid frisked searched contrabn  \
0  TRESPAS        0       0       N                Y        Y        N   
1                 0       0       N                N        Y        N   
2                 0       0       Y                Y        Y        N   
3                 0       0       Y                Y        N        N   
4                 0       0       Y                N        N        N   

  adtlrept pistol riflshot asltweap knifcuti machgun othrweap pf_hands  \
0        N      N        N        N        N       N        N        N   
1        N      N        N        N        N       N        N        Y   
2        N      N        N        N        N       N        N        N   
3        N      N        N        N        N       N        N        Y   
4        N      N        N        N        N       N        N        N   

  pf_wall pf_grnd pf_drwep pf_ptwep pf_baton pf_hcuff pf_pepsp pf_other radio  \
0       N       N        N        N        N        Y        N        N     N   
1       Y       N        N        N        N        N        N        N     Y   
2       N       N        N        N        N        N        N        N     Y   
3       N       N        N        Y        N        Y        N        N     N   
4       N       N        N        N        N        N        N        N     N   

  ac_rept ac_inves rf_vcrim rf_othsw ac_proxm rf_attir cs_objcs cs_descr  \
0       N        N        N        N        N        N        N        N   
1       N        N        N        N        N        N        N        N   
2       Y        N        N        Y        N        N        N        Y   
3       N        N        Y        N        Y        N        N        Y   
4       N        N        N        N        N        N        N        N   

  cs_casng cs_lkout rf_vcact cs_cloth cs_drgtr ac_evasv ac_assoc cs_furtv  \
0        Y        N        N        N        N        N        Y        Y   
1        N        N        N        N        Y        Y        N        Y   
2        N        N        N        N        N        N        N        N   
3        N        N        N        N        N        N        N        Y   
4        N        N        N        N        N        N        N        N   

  rf_rfcmp ac_cgdir rf_verbl cs_vcrim cs_bulge cs_other ac_incid ac_time  \
0        N        Y        N        N        N        N        Y       Y   
1        N        N        N        N        N        Y        Y       Y   
2        N        N        N        N        N        N        N       N   
3        Y        Y        N        N        N        N        N       N   
4        N        N        N        N        N        Y        N       N   

  rf_knowl ac_stsnd ac_other sb_hdobj sb_outln sb_admis sb_other repcmd  \
0        N        N        N        N        N        N        Y     41   
1        N        N        N        N        N        N        Y     10   
2        N        N        N        N        N        N        Y     66   
3        N        N        N        N        N       

## Summary statistics

Show some summary statistics :

### <span style='color:green'> Share of people arrested $=>$ clean this variable</span>

There seeems to be a missing variable : remove the observation 

### <span style='color:green'>Distribution of race</span>

### <span style='color:green'>Number and % of people arrested and % of people not arrested within each racial category.</span>

In other words, of the people that are arrested, what percent are Black, White, Hispanic etc. Of the people that are not arrested, what percent are Black, White, Hispanic etc..

Use the `groupby` syntax from `pandas`.

**Percentage of people from different racial category, for the frisked group and non-frisked group, respectively**

## Visualization

### <span style='color:green'>Plot the distribution of race among stopped individual</span>

### Create a plot of arrests made vs. race 

#### <span style='color:green'>Create a variable `arst_dummy` equal to 1 if the stop led to an arrest and 0 if not.</span>

You can use the `apply` syntax from Pandas. 

#### <span style='color:green'>Represent the previous summary statistics for the percent of people within each racial group that were arrested  with a bar plot instead.</span>

The bar plot should have the title "Percent of Racial Group Arrested".
Now you can plot arrests made vs. race using the previous dummy. 
You can use the `barplot` function from `seaborn`.

### What is the distrbution of ages by those who are arrested vs. not?

#### <span style='color:green'>Cleaning step: convert `age` to a numeric</span>

#### <span style='color:green'>Plot the age distribution by those who are arrested vs. not</span>

You can use `displot` from seaborn or `hist` from matplotlib 

### Writing function & using loops

Many of the variables in the stop and frisk data are coded as "Y" for "Yes" and "N" for no. 

#### <span style='color:green'>Propose an easy means of recoding every variable in the stop and frisk data set using a function that you define. </span>

In order to save some time from having to recode every single variable that contains a "Y" or a "N", write a function that transforms:

* "Y" codings to 1
* "N" codings to 0
* " " codings to np.NAN (missing)

You should be able to use this function in an `apply` framework.

#### <span style='color:green'>`for` loop </span>

Using the `yesno` function, write a loop that transforms every single variable in the "stopandfrisk2016" data frame containing a "Y" or "N" coding into "1", "0" or "NA" codings as specified above. 

Save these newly coded variables in a data frame called `recoded` and use the `head()` function to print out the first few observations of the new dataframe that you created.